### ISSUE TITLE
DMA Handle writing data where source data is not in RAM

### ISSUE THREAD

**Issue by MabezDev** [esp-rs/esp-hal#674]
Here is a demo of the problem: https://wokwi.com/projects/370955476071667713

You can remove the call to `with_dma` and the example works. This might be because transferring such a small number of bytes with DMA isn't possible, but I didn't see anything in the trm about this limitation, so it should just work.

**Comment by MabezDev** [comment]
Weird but it looks like the operation at least tries to go, the clock signal runs correctly, but MOSI is just firing out zeroes. Almost like its not actually sending the data we're telling it to, but something else.

![image](https://github.com/esp-rs/esp-hal/assets/6977289/93d2660e-a8e7-48e1-9fd7-abd2ae391199)

**Comment by MabezDev** [comment]
From 2.4.5 of the esp32c3 trm:
> Buffer address pointer (DW1) check. If the buffer address pointer points to 0x3FC80000 ~ 0x3FCDFFFF
(please refer to Section 2.4.7), it passes the check.

So I guess if we try and send data that might be in DROM i.e [some const array](https://github.com/rust-embedded-community/embedded-sdmmc-rs/blob/ea53bc97f9d0ebab078360425f058509cb15585b/src/sdcard/mod.rs#L400) it won't work :/. I'll double-check that this is what's going on, but this is my suspicion.

Even if it isn't, its something we need to handle, I can think of three options:

1) Error out (I think with #665 this should be the case?) and make the user deal with it
2) Keep around a static buffer for the device and whenever we encounter ROM data, copy into RAM first and then send it
3) Switch to non-dma spi mode for these transfers, copy via CPU & SPI FIFOs

Because its hard to control where consts are placed (especially if they're embedded inside a 3rd party crate), we have to implement either 2 or 3 here, IMO. I would be interested to hear others thoughts, in particular @bjoernQ.

**Comment by bugadani** [comment]
This is somewhat similar to nRF52's EasyDma: IIRC it can only access the RAM, not flash. It might be worth checking how this issue is dealt with there.

**Comment by MabezDev** [comment]
So testing on real hardware using #665 I get the following output:

```rust
TRACE - Reset card..
INFO - Address of out buffer: 0x3C010454
0x3C010454 - .Lanon.f4747abe48f02cefea07fa6b9da986da.24
    at ??:??
ERROR - Transport error: DmaError(DescriptorError)
```

#665 is a step in the right direction, at least now it doesn't hang, but obviously we need to solve this.

> This is somewhat similar to nRF52's EasyDma: IIRC it can only access the RAM, not flash. It might be worth checking how this issue is dealt with there.

Thanks! I'll take a look there.

**Comment by MabezDev** [comment]
Looks like in embassy, they just report an error: https://github.com/embassy-rs/embassy/blob/4db63677f6e56c52ea3d432542513967f5d93c69/embassy-nrf/src/util.rs#L24

**Comment by MabezDev** [comment]
> Looks like in embassy, they just report an error: https://github.com/embassy-rs/embassy/blob/4db63677f6e56c52ea3d432542513967f5d93c69/embassy-nrf/src/util.rs#L24

Ah looks like they handle this in the actual impls by copying into a buffer: https://github.com/embassy-rs/embassy/blob/4db63677f6e56c52ea3d432542513967f5d93c69/embassy-nrf/src/spis.rs#L277-L285

**Comment by MabezDev** [comment]
This is also the case in esp-idf, although they have a heap and can alloc the perfect sized array. We will have to pick a size at compiler time (and be careful not to stack allocate huge arrays).

**Comment by MabezDev** [comment]
I have proposed two possible solutions in #677 and #678, I'd appreciate some thoughts on the ideas presented in those PRs. I think if we could find a clean way to make it seamless (#677) I would prefer that, but right now I'm not seeing how its possible.

**Comment by bjoernQ** [comment]
To be honest when I implemented this, I wasn't sure if it's a good idea to implement the EH traits for a DMA-enabled SPI instance at all - and maybe it was a bad idea to do it

My concerns are EH doesn't state any requirements about the buffers but when using DMA there are such requirements.

What I mean is that in theory we should be able to use any code working with the EH traits and feed it a DMA-enabled SPI instance and it should work.

I'm not sure if we really need an explicit check on the memory region since I guess the hardware detects the problem already (seems like Wokwi doesn't -but it should)

**Comment by MabezDev** [comment]
> To be honest when I implemented this, I wasn't sure if it's a good idea to implement the EH traits for a DMA-enabled SPI instance at all - and maybe it was a bad idea to do it
> 
> My concerns are EH doesn't state any requirements about the buffers but when using DMA there are such requirements.
> 
> What I mean is that in theory we should be able to use any code working with the EH traits and feed it a DMA-enabled SPI instance and it should work.
> 
> I'm not sure if we really need an explicit check on the memory region since I guess the hardware detects the problem already (seems like Wokwi doesn't -but it should)

Even if it is a bad idea to impl the blocking traits for SpiDma (which I personally don't think was a bad idea) we'd still run into this issue for the async traits, as they pretty much rely on the DMA for their asynchronousness.

The fundamental problem is that whether something is stored on the stack or in rodata is actually optimization dependant sometimes, this means even if users are careful they can still run into this issue. If the issue occurs within a driver, then they are stuck. I think we have to handle this in some capacity, whether its the two approaches provided above, or a third.

Another thought I had is defaulting to using SPI + FIFO (CPU) instead of DMA if the DMA is incapable of sending a specific buffer.

**Comment by MabezDev** [comment]
Closed via #678
